File: MPP/MPP Proiect/CSharp/protobuf/ProtoWorker.cs
using System;
using System.Net.Sockets;
using System.Threading;
using Google.Protobuf;
using Java;
using services;

namespace protobuf{
    public class ProtoWorker: IObserver{
        private IServices server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;

        public ProtoWorker(IServices server, TcpClient connection){
            this.server = server;
            this.connection = connection;
            try{
                stream = connection.GetStream();
                connected = true;
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void Run(){
            while(connected){
                try{
                    Request request = Request.Parser.ParseDelimitedFrom(stream);
                    Response response = handleRequest(request);
                    if(response != null){
                        sendResponse(response);
                    }
                } catch(Exception e){
                    Console.WriteLine(e.StackTrace);
                }

                try{
                    Thread.Sleep(1000);
                } catch(Exception e){
                    Console.WriteLine(e.StackTrace);
                }
            }

            try{
                stream.Close();
                connection.Close();
            } catch(Exception e){
                Console.WriteLine("Error " + e);
            }
        }

        private Response handleRequest(Request request){
            Response response = null;
            Request.Types.Type reqType = request.Type;
            switch(reqType){
                case Request.Types.Type.Login:{
                    Console.WriteLine("Login request ...");
                    model.Referee referee = ProtoUtils.GetReferee(request);
                    try{
                        lock(server){
                            return ProtoUtils.CreateLoginResponse(
                                server.Login(referee, this));
                        }
                    } catch(MyException e){
                        connected = false;
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }
                }
                case Request.Types.Type.Logout:{
                    Console.WriteLine("Logout request");
                    model.Referee referee = ProtoUtils.GetReferee(request);
                    try{
                        lock(server){
                            server.Logout(referee);
                        }

                        connected = false;
                        return ProtoUtils.CreateOkResponse();
                    } catch(MyException e){
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }
                }
                case Request.Types.Type.AddTest:{
                    Console.WriteLine("Add test request");
                    model.Test test = ProtoUtils.GetTest(request);
                    try{
                        lock(server){
                            server.AddTest(test);
                        }

                        return ProtoUtils.CreateOkResponse();
                    } catch(MyException e){
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }
                }
                case Request.Types.Type.GetAllParticipants:{
                    Console.WriteLine("Get all participants request");
                    try{
                        lock(server){
                            return ProtoUtils.CreateGetAllParticipantsResponse(
                                server.GetAllParticipantsSortedByName());
                        }
                    } catch(MyException e){
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }
                }
                case Request.Types.Type.GetAllActiveParticipants:{
                    Console.WriteLine("Get all active participants request");
                    try{
                        var id = ProtoUtils.GetId(request);
                        lock(server){
                            return ProtoUtils.CreateGetAllActiveParticipantsResponse(
                                server.GetAllActiveParticipants(id));
                        }
                    } catch(MyException e){
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }
                }
            }

            return response;
        }

        private void sendResponse(Response response){
            Console.WriteLine("sending response " + response);
            lock(stream){
                response.WriteDelimitedTo(stream);
                stream.Flush();
            }
        }

        public void RefreshAll(model.Test test){
            Console.WriteLine("Test added " + test);
            try{
                sendResponse(ProtoUtils.CreateTestAddedResponse(test));
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }
    }
}

File: MPP/MPP Proiect/Java/Networking/src/main/java/protobuffprotocol/ProtoWorker.java
package protobuffprotocol;

import model.Participant;
import model.Referee;
import model.Test;
import service.IObserver;
import service.IService;
import service.MyException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;

public class ProtoWorker implements Runnable, IObserver{
    private final IService server;
    private final Socket connection;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ProtoWorker(IService server, Socket connection){
        this.server = server;
        this.connection = connection;
        try{
            output = connection.getOutputStream();
            input = connection.getInputStream();
            connected = true;
        }catch(IOException e){
            e.printStackTrace();
        }
    }

    @Override
    public void run(){
        while(connected){
            try{
                //Object request=input.readObject();
                System.out.println("Waiting requests ...");
                Protobufs.Request request = Protobufs.Request.parseDelimitedFrom(input);
                System.out.println("Request received: " + request);
                Protobufs.Response response = handleRequest(request);
                if(response != null){
                    sendResponse(response);
                }
            }catch(IOException e){
                e.printStackTrace();
            }
            try{
                Thread.sleep(1000);
            }catch(InterruptedException e){
                e.printStackTrace();
            }
        }
        try{
            input.close();
            output.close();
            connection.close();
        }catch(IOException e){
            System.out.println("Error " + e);
        }
    }

    @Override
    public void refreshAll(Test test){
        System.out.println("RefreshAll ... with test: " + test);
        try{
            sendResponse(ProtoUtils.createRefreshAllResponse(test));
        }catch(IOException e){
            e.printStackTrace();
        }
    }

    private Protobufs.Response handleRequest(Protobufs.Request request){
        switch(request.getType()){
            case LOGIN:{
                System.out.println("Login request ...");
                Referee referee = ProtoUtils.getReferee(request);
                try{
                    Referee loggedReferee = server.login(referee, this);
                    return ProtoUtils.createOkResponse(loggedReferee);
                }catch(MyException e){
                    connected = false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case LOGOUT:{
                System.out.println("Logout request");
                Referee referee = ProtoUtils.getReferee(request);
                try{
                    server.logout(referee, this);
                    connected = false;
                    return ProtoUtils.createOkResponse();

                }catch(MyException e){
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case ADD_TEST:{
                System.out.println("ADD_TEST Request ...");
                Test test = ProtoUtils.getTest(request);
                try{
                    server.addTest(test);
                    return ProtoUtils.createOkResponse(test);
                }catch(MyException e){
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GET_ALL_PARTICIPANTS:{
                System.out.println("Get_All_Participants Request ...");
                try{
                    ArrayList<Participant> participants =
                            (ArrayList<Participant>) server.getAllParticipants();
                    return ProtoUtils.createGetAllParticipantsResponse(participants);
                }catch(MyException e){
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            }
            case GET_ALL_ACTIVE_PARTICIPANTS:{
                System.out.println("Get_All_Active_Participants Request ...");
                int id = ProtoUtils.getId(request);
                try{
                    ArrayList<Participant> activeParticipants =
                            (ArrayList<Participant>) server.getAllActiveParticipants(id);
                    return ProtoUtils.createGetAllActiveParticipantsResponse(activeParticipants);
                }catch(MyException e){
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
        }
        return null;
    }

    private void sendResponse(Protobufs.Response response) throws IOException{
        System.out.println("sending response " + response);
        response.writeDelimitedTo(output);
        //output.writeObject(response);
        output.flush();
    }
}
